File: LCCI/1-5.one-away-lcci/one-away-lcci.java
class Solution {
    public boolean oneEditAway(String first, String second) {
        if (first == "" && second == "") {
            return true;
        }
        if (first.length() == second.length()) {
            return oneEditReplace(first,second);
        } else if (first.length() + 1 == second.length()) {
            return oneEditInsert(first,second);
        } else if (first.length() - 1 == second.length()) {
            return oneEditInsert(first,second);
        }
        return false;
    }

    public boolean oneEditReplace(String s1,String s2) {
        boolean foundDifference = false;
        for (int i = 0;i < s1.length();i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                if (foundDifference) {
                    return true;
                }
                foundDifference = false;
            } 
        }
        return false;
    }

    public boolean oneEditInsert(String s1,String s2) {
        int index1 = 0;
        int index2 = 0;
        while (index2 < s2.length() && index1 < s1.length()) {
            if (s1.charAt(index1) != s2.charAt(index2)) {
                if (index1 != index2) {
                    return false;
                }
                index2++;
            } else {
                index1++;
                index2++;
            }
        }
        return true;   
    }
}

File: LCCI/1-5.one-away-lcci/one-away-lcci.go
func oneEditAway(first string, second string) bool {
    if math.Abs(float64(len(first))-float64(len(second))) > 1 { //长度相差超过1就返回
		return false
	}
	length1, length2 := len(first), len(second)
	differ := 0 //记录两个字符串比较有多少次的不同
	i, j := 0, 0
	for i < length1 && j < length2 {
		if first[i] != second[j] {
			differ++
			if length1 > length2 { //出现不同的情况，就把长的那个下标++，正常来说，下一个会相等，并且之后的字符也会相等
				i++
				continue
			} else if length1 < length2 {
				j++
				continue
			}
		}
		i++
		j++
	}

	return differ <= 1
}